core/pkg/service/sync/server.go

```diff
@@ -0,0 +1,106 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	rpc "buf.build/gen/go/open-feature/flagd/bufbuild/connect-go/sync/v1/syncv1connect"
+	"github.com/open-feature/flagd/core/pkg/logger"
+	iservice "github.com/open-feature/flagd/core/pkg/service"
+	syncStore "github.com/open-feature/flagd/core/pkg/sync-store"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+)
+
+type Server struct {
+	server  http.Server
+	Logger  *logger.Logger
+	handler handler
+	config  iservice.Configuration
+}
+
+func NewServer(ctx context.Context, logger *logger.Logger) *Server {
+	syncStore := syncStore.NewSyncStore(ctx, logger)
+	return &Server{
+		handler: handler{
+			logger:    logger,
+			syncStore: syncStore,
+		},
+		Logger: logger,
+	}
+}
+
+func (s *Server) Serve(ctx context.Context, svcConf iservice.Configuration) error {
+	s.config = svcConf
+	lis, err := s.setupServer()
+	if err != nil {
+		return err
+	}
+
+	go s.bindMetrics()
+
+	errChan := make(chan error, 1)
+	go func() {
+		if err := s.server.Serve(
+			lis,
+		); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errChan <- err
+		}
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return s.server.Shutdown(ctx)
+	}
```

Comment:
Potentially a refactoring suggestion - can be a follow-up 

Here, we start two servers (listeners) 1. Metrics 2. GRPC streams

Both can be wrapped in errgroup to handle errors and avoid the error chan. Further,  setupServer() is not needed in top level if we do this . Psecudo code look like,

```
group, gCtx := errgroup.WithContext(ctx)

group.go(startGrpcServer)

group.go(startMetricsServer)

group.wait()
```

This makes it easy to read the code as well as test inidividual server impls :)

Reply:
nice suggestion, ill implement this tomorrow

Reply:
done